Propose Github actions for CI

tristanbes opened the pull request:
Github actions is very reliable and fast for a CI use;

I've configured it to run:

- phpstan
- phpunit
- php-cs-fixer
- *phpcs*
- composer validate
- validation of twig files
- validation of yaml files
- database schema validation (currently broken because it seems there is a migration file missing, fixed in this PR)

```
bin/console do:sc:up --dump-sql

 The following SQL statements will be executed:

     ALTER TABLE commissionable_asset ALTER has_mobile_radio SET NOT NULL;
```

You can see full output there: 
https://github.com/tristanbes/resop/pull/1/checks

The matcher will report the error inside the pull request on the "code" tab;

### Results

https://github.com/tristanbes/resop/pull/1

![propose_github_actions_for_CI_by_tristanbes_·_Pull_Request__1_·_tristanbes_resop](https://user-images.githubusercontent.com/346010/77934509-7c184f80-72b0-11ea-9309-eac18faa41d8.png)

mRoca commented:
Using Github actions would be great, indeed! But are you sure there is no way to use our docker images ? As you can see in https://gitlab.com/mRoca/resop/pipelines , the jobs are quite fast and only use the docker stack that every dev is already using on this project : https://github.com/crf-devs/resop/blob/master/.gitlab-ci.yml#L33-L53

tristanbes commented:
@mRoca I have 0 knowledge width Docker; I'm sure it's possible but have no clue nor intention to work with docker;

I'm offering a working solution for now that we use on a lot of repository at the company I work for; Feel free to merge it to improve contributions from other developpers and maybe work on a Docker solution later ? What do you think ?

mRoca commented:
Ok ! I think we'll merge your PR, it's a really good idea :-) Let me check for this Docker question tonight

tristanbes commented:
I've added `phpcs`, tweaked the rules (minimum impact) and fixed all the errors; Now, all quality toolset is present and correctly configured + checked with Github Actions ;)

mRoca commented on .php_cs.dist:
```diff
@@ -1,15 +1,27 @@
         'declare_strict_types' => true,
+        'braces'                                => ['allow_single_line_closure' => true],
+        'native_constant_invocation'            => true,
+        'native_function_invocation'            => [
```
why not using `@Symfony:risky` instead of all those rules ?

tristanbes replied:
Not all the rules are contained inside Symfony:risky as far as i've checked, and Symfony:risky contains way more rules that the ones i've added;

mRoca commented on composer.json:
```diff
@@ -1,6 +1,8 @@
   "type": "project",
   "license": "proprietary",
+  "name": "crf-devs/resop",
+  "description": "resop Croix Rouge Française",
```
`ResOp`

mRoca replied:
and you should remove ` Croix Rouge Française` for now

mRoca commented on src/Domain/PlanningUpdateDomain.php:
```diff
@@ -116,23 +117,23 @@ protected function process(string $entityClass, string $availabilityClass, array
             }
 
-            /** @var AvailabilityInterface */
+            /* * @var AvailabilityInterface */
```
why ?

tristanbes replied:
It's a mistake; i've fixed it.

mRoca commented on src/Migrations/Version20200330155339.php:
```diff
@@ -0,0 +1,36 @@
+        // this down() migration is auto-generated, please modify it to your needs
+        $this->abortIf('postgresql' !== $this->connection->getDatabasePlatform()->getName(), 'Migration can only be executed safely on \'postgresql\'.');
+
+        $this->addSql('CREATE SCHEMA public');
```
to remove

mRoca commented on src/Migrations/Version20200330155339.php:
```diff
@@ -0,0 +1,36 @@
+
+    public function down(Schema $schema): void
+    {
+        // this down() migration is auto-generated, please modify it to your needs
```
to remove

mRoca commented on src/Migrations/Version20200330155339.php:
```diff
@@ -0,0 +1,36 @@
+
+    public function up(Schema $schema): void
+    {
+        // this up() migration is auto-generated, please modify it to your needs
```
to remove

mRoca commented on src/Migrations/Version20200330155339.php:
```diff
@@ -0,0 +1,36 @@
+{
+    public function getDescription(): string
+    {
+        return '';
```
can you add a description ?

tristanbes replied:
Someone forgot to generate this migration, so I can't invent one; 
Also adding a description only to say "add a field X" when you can read it from the SQL statement is purely useless.

mRoca replied:
Then you can remove the method, as it's the same as the parent's one

mRoca commented on src/Migrations/Version20200330155339.php:
```diff
@@ -0,0 +1,36 @@
+use Doctrine\Migrations\AbstractMigration;
+
+/**
+ * Auto-generated Migration: Please modify to your needs!
```
to remove

mRoca commented on composer.json:
```diff
@@ -31,11 +33,14 @@
     "phpstan/phpstan-symfony": "^0.12.2",
     "phpunit/phpunit": "^7.5",
     "roave/security-advisories": "dev-master",
+    "squizlabs/php_codesniffer": "^3.5",
```
if you add csniffer, can you add the related `test` and `fix-cs` commands in the Makefile ?

tristanbes replied:
sure

mRoca commented:
I've checked your pipelines, is there any way to get the commands logs / output ?

mRoca commented:
Please don't add more tools in this PR: it's big enough for now :-) We can add them after merging this one, and updating all project files in a dedicated PR can be easier to deal with

tristanbes commented:
CR done:

The output will be directly inside the modified code on the "File changed" view; which is way better than parsing an output with your eyes since it's contextualized.

Also: Github actions always outputs the output where there is one. When you see none, it means status code exit 0, which means no errors found when it comes to CS.

For other check, some commands are more verbose than others, let's take for example this one: https://github.com/tristanbes/resop/pull/1/checks?check_run_id=546502893#step:19:17

mRoca commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,128 @@
+                  php-version: ${{ env.PHP_VERSION }}
+                  coverage: none
+                  extensions: ctype, iconv, intl
+                  ini-values: date.timezone=Europe/Paris, memory_limit=-1, session.gc_probability=0, opcache.enable=1, opcache.enable_cli=1, opcache.memory_consumption=256, opcache.max_accelerated_files=20000, opcache.validate_timestamps=0, realpath_cache_size=4096K, realpath_cache_ttl=600
```
UTC

mRoca commented on Makefile:
```diff
@@ -92,22 +92,24 @@ fix-cs: fix-cs-php
+	bin/tools bin/php-cs-fixer fix --allow-risky yes --dry-run --verbose --diff
+	bin/tools bin/phpcs
+	bin/tools bin/console lint:twig templates --env=test
+	bin/tools php bin/console lint:yaml config --parse-tags --env=test
```
why `php` ?

tristanbes replied:
fixed

mRoca commented on Makefile:
```diff
@@ -79,7 +79,7 @@ init-db: start-db
 	bin/tools bin/post-install-dev.sh
 
 fix-cs-php:
```
can't you add https://github.com/squizlabs/PHP_CodeSniffer/wiki/Fixing-Errors-Automatically  here ?

tristanbes replied:
done

mRoca commented on Makefile:
```diff
@@ -92,22 +92,24 @@ fix-cs: fix-cs-php
 test-advanced:
 	bin/tools sh -c "APP_DEBUG=1 APP_ENV=test bin/console c:w"
-	bin/tools vendor/bin/phpstan analyse --configuration phpstan.neon --level max --no-progress .
+	bin/tools bin/phpstan.phar analyse --configuration phpstan.neon --level max --no-progress .
```
why do you prefer the phar ?

tristanbes replied:
because it's budled, it doesn't have any depandancy. but in this particular situation, it doesn't make any difference because it's a dev requirement.

mRoca commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,128 @@
+              run: php bin/phpstan.phar analyse --configuration phpstan.neon --level max --no-progress . --error-format=checkstyle | cs2pr
+              if: always()
+
+            - run: bin/phpunit --printer mheap\\GithubActionsReporter\\Printer
```
Can't we keep the phpdbg code coverage report ?

tristanbes replied:
I don't know; how was it generated before ?
Also, are you reffering to https://github.com/krakjoe/phpdbg this ? because it's archived and unmaintained for 6 years 😱

tristanbes replied:
Also: the action i'm using to setup PHP can support code coverage using Xdebug or PCOV (which is faster); see : https://github.com/marketplace/actions/setup-php-action#signal_strength-coverage-support

Maybe re-enable code coverage in another PR ?

mRoca replied:
We'll re-enable code coverage later, ok. But you should check what was existing before creating a PR updating everything, don't you think ? ;-) 

https://gitlab.com/mRoca/resop/-/jobs/491684797 > you can see the code coverage report. I've removed the coverage badge from the README only because there is a but on displaying it on github, you can see it there : https://gitlab.com/mRoca/resop/-/tree/2fc11fd3b22834772596da8f36f008ba7f3bf198#contributing

And for your information, the krakjoe/phpdbg has been archived 6 years ago because phpdbg has been added to the php-src repository itself, from php5.6.

tristanbes replied:
Sure, you got a point; I'd assume you had no CI since the badge were broken, and even the link you provided lands on 404; My bad :)

mRoca commented on composer.json:
```diff
@@ -44,7 +49,8 @@
     },
-    "sort-packages": true
+    "sort-packages": true,
+    "bin-dir": "bin/"
```
is there any reason to do it ? As we are always using command runners (Makefile, GithubActions, ...), are you sure we need those bins in the root bin dir ?

tristanbes replied:
It doesn't really matter because they're ignored (only bin/console and user generated bin are commited). I like to have all the binaries inside the global bin/ directory. IT's messy when you try to manually launch commands and try to remember if you need to vendor/bin versus bin/

mRoca replied:
As you are working on an already existing project, I'm not sure you'll always have what you like to have in an open source project with many other people. For my part I would prefer to have vendor related bin files in the vendors folder, and app related bin files in the root folder. And other contributor can prefer other things.

Some points :
- You PR is really valuable: it should be nice to use Github Actions in a github hosted project!
- We should keep in mind that our current GitlabCI config was doing exactly the same job in 10 lines instead of 128.
- Github actions are better than GitlabCI for a github project because of one point : GitlabCI cannot deal with Pull Request from another repository
- Adding phpcs and phpcbf is a really good idea, but maybe we should have added it in a dedicated PR in order to merge this one as soon as possible
- Removing code coverage is not a very good thing, but as the current coverage is about 7%, it's not so important
- Same thing for changing the php-cs-fixer config: imho, a dedicated PR would have been easier to review
- I don't see the point of using .phar file and another bin-dir if there is no need and if it's not something strongly recommended

=> Could you revert the not so useful changes like .phar and bin-dir ? We can keep everything else and see later for those points !

tristanbes replied:
Sure, I'll revert it.
Your logic is broken for bin/console:  it's a vendor one and is not inside the vendor ;)

mRoca commented:
Looks almost good for me, some more comments  but we're almost done!

mRoca commented:
LGFM, thank you very much, @tristanbes ! Can I merge it ?

tristanbes commented:
Yes you can, it's all green there:
https://github.com/tristanbes/resop/pull/1/checks?check_run_id=549841026
https://github.com/tristanbes/resop/pull/1/checks?check_run_id=549841007

👍
